### ISSUE TITLE
semicolon-if-nothing-returned false positive with tokio macro

### ISSUE THREAD

**Issue by flisky** [rust-lang/rust-clippy#7438]
Lint name: semicolon-if-nothing-returned

I tried this code:

```rust
#[tokio::test]
async fn test_ok() {
    assert!(true);
}
```

Instead, this happened: 

```
   |
nn |         assert!(true);
   |         ^^^^^^^^^^^^^^ help: add a `;` here: `assert!(true);;`
   |
   = note: `-D clippy::semicolon-if-nothing-returned` implied by `-D clippy::pedantic`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned
```
### Meta

- `cargo clippy -V`: clippy 0.1.55 (952fdf2a 2021-07-05)
- `rustc -Vv`:
```
rustc 1.55.0-nightly (952fdf2a1 2021-07-05)
binary: rustc
commit-hash: 952fdf2a1119affa1b37bcacb0c49cf9f0168ac8
commit-date: 2021-07-05
host: x86_64-apple-darwin
release: 1.55.0-nightly
LLVM version: 12.0.1
```

**Comment by giraffate** [comment]
I tried it on the [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8cd70b345024cacab08111c4d5c3305b), but couldn't reproduce that warning. Can you provide reproduction on the playground?

**Comment by flisky** [comment]
Sorry, I cannot reproduce this on playground, either. However, it's reproduced on a simple cargo project -

* create a project: `cargo new semicolon`
* add tokio dependency: `tokio = { version = "*", features = ["macros", "rt-multi-thread", "time"] }`
* add the following to `main.rs`:

```
#![warn(clippy::semicolon_if_nothing_returned)]

#[tokio::main]
async fn main() {
    assert!(true);
}
```

output:
```
warning: consider adding a `;` to the last statement for consistent formatting
 --> src/main.rs:5:5
  |
5 |     assert!(true);
  |     ^^^^^^^^^^^^^^ help: add a `;` here: `assert!(true);;`
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::semicolon_if_nothing_returned)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned

warning: `assert!(true)` will be optimized out by the compiler
 --> src/main.rs:5:5
  |
5 |     assert!(true);
  |     ^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::assertions_on_constants)]` on by default
  = help: remove it
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assertions_on_constants
  = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: 2 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1.11s
```

**Comment by giraffate** [comment]
Thanks, I can reproduce it!
```
$ rustc -V
rustc 1.55.0-nightly (7a16cfcff 2021-07-11)
$ cargo clippy -V
clippy 0.1.55 (7a16cfcf 2021-07-11)
```

**Comment by giraffate** [comment]
Reproduction on the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ca32bde812b8341a2545a9e64b40a58d).

**Comment by c0va23** [comment]
I face with same issue when update `tokio-macros` from version 1.1.0 to 1.2.0 or 1.3.0.
```
$ cargo clippy --version
clippy 0.1.54 (a178d03 2021-07-26)
```

**Comment by c0va23** [comment]
The problem occurs because the macros 'tokio:: main` and' tokio::test` wrap functions and the last expressions are always not terminated with a semicolon.

I will try to fix this problem on the `tokio-macros ' side (tokio-rs/tokio#4067), but this problem is not specific to tokio. It looks like a fix on the clippy side is also necessary.
